### ISSUE TITLE
[Flang] Incorrect diagnostic on sequence association of character type arguments

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#123807]
Consider the following code
```
module m
    contains

    subroutine test2 (s)
        character(5), intent(inout) :: s(5)

    !    s = (/character(5) :: 'POINT', 'COMMA', 'UNDEFINED', '',''/)
    end subroutine

end module

use m
    character(30), allocatable :: s3

    allocate (s3)
    s3 = ''

    call test2(s3)
    end
```

Flang currently issues an error as:
```
./t.f:18:16: error: Actual argument has fewer characters remaining in storage sequence (24) than dummy argument 's=' (25)
      call test2(s3)
                 ^^
```

This is sequence association. It is allowed to have different length between actual and dummy character type argument.
All ifort, gfortran and XLF compiled the code successfully.

**Comment by klausler** [comment]
F'2023 15.5.2.12 ("Sequence association") paragraph 6: "The rank and shape of the actual argument need not agree with the rank and shape of the dummy argument, but the number of elements in the dummy argument shall not exceed the number of elements in the element sequence of the actual argument."

**Comment by DanielCChen** [comment]
Thanks.
Just out of curiosity, there is number `24` coming from in the error message?

**Comment by klausler** [comment]
> Thanks. Just out of curiosity, there is number `24` coming from in the error message?

That looks like a bug; will fix.

**Comment by DanielCChen** [comment]
I commented out the assignment that sets `s3` to zero length, so it is now a character string with length of 30, but Flang stills complains the same error.

```
module m
    contains

    subroutine test2 (s)
        character(5), intent(inout) :: s(5)

    !    s = (/character(5) :: 'POINT', 'COMMA', 'UNDEFINED', '',''/)
    end subroutine

end module

use m
    character(30), allocatable :: s3

    allocate (s3)
!    s3 = ''

    call test2(s3)
    end
```

**Comment by klausler** [comment]
The type is `character(30)` and will always have length 30.  Reallocation to a new length on assignment only happens with deferred-length allocatables.

**Comment by DanielCChen** [comment]
Right. Compiling the updated code will still trigger the same error
```
error: Semantic errors in 123807.f
./123807.f:18:16: error: Actual argument has fewer characters remaining in storage sequence (24) than dummy argument 's=' (25)
      call test2(s3)
                 ^^
```

Now the number of elements in dummy argument (25) does not exceed the number of element in the actual actual argument (30), so the sequence association should be conforming now.

**Comment by DanielCChen** [comment]
After reading the standard again, I think the original code is conforming.
The standard says:
```
The element sequence is the sequence of consecutive groups of storage units in the storage unit sequence, grouped by the character length of the dummy array. The sequence terminates when the number of storage units left is less than the character length of the dummy array.
```

This makes the actual argument to have 6 elements in the element sequence (30/5 = 6).
It makes the dummy argument to have 5 elements in the element sequence.

Then
```
The rank and shape of the actual argument need not agree with the rank and shape of the dummy argument, but the number of elements in the dummy argument shall not exceed the number of elements in the element sequence of the actual argument.
```
